File: python/row_manager.py
from shapes import Row


class RowManager:
    def __init__(self, current_width=1, grid_width=10, offset=10):
        self.current_row = 1
        self.grid_width = grid_width
        self.offset = offset
        self.current_width = current_width
        self.positions = self.calculate_x_positions()
        self.rows = [Row(self.current_row, self.positions)]

    def calculate_x_positions(self):
        positions = list(range(self.current_width))
        positions = [self.offset + p - self.current_width / 2 for p in positions]
        positions = [p * self.grid_width + self.offset for p in positions]
        return positions

    def add_row(self):
        xpos = self.calculate_x_positions()
        self.current_row += 1
        self.rows.append(Row(self.current_row, xpos))

    def remove_row(self):
        self.current_row -= 1
        self.rows.pop()

    def next_row(self, color):
        xpos = self.calculate_x_positions()
        return Row(self.current_row + 1, xpos, color=color)

    def get_all_circles(self):
        all_circles = []
        for row in self.rows:
            all_circles.extend(row.get_circles())
        return all_circles

File: app/simulation/row-manager.ts
import { Circle, Row } from "./shapes";
import { Color } from "./types";

export class RowManager {
  constructor(
    public currentWidth = 1,
    private gridWidth = 10,
    private offset = 10,
    private currentRow = 1,
    private positions: number[] = [],
    public rows: Row[] = []
  ) {
    this.positions = this.calculateXPositions();
    this.rows = [new Row(this.currentRow, this.positions)];
  }

  private calculateXPositions() {
    return Array.from({ length: this.currentWidth })
      .map((_, i) => i)
      .map((i) => this.offset + i - this.currentWidth / 2)
      .map((i) => i * this.gridWidth + this.offset);
  }

  public addRow() {
    const xPositions = this.calculateXPositions();
    this.currentRow++;
    this.rows.push(new Row(this.currentRow, xPositions));
  }

  public removeRow() {
    this.currentRow -= 1;
    this.rows.pop();
  }

  public nextRow(color: number | Color) {
    const xPositions = this.calculateXPositions();
    return new Row(this.currentRow + 1, xPositions, undefined, color);
  }

  public getAllCircles() {
    const allCircles: Circle[] = [];

    this.rows.forEach((row) => {
      allCircles.push(...row.getCircles());
    });

    return allCircles;
  }
}
